src/Orleans/Configuration/GlobalConfiguration.cs

```diff
@@ -670,13 +671,83 @@ internal override void Load(XmlElement root)
                         {
                             var providerConfig = new ProviderCategoryConfiguration();
                             providerConfig.Load(child);
-                            ProviderConfigurations.Add(providerConfig.Name, providerConfig);
+
+                            if (ProviderConfigurations.ContainsKey(providerConfig.Name))
+                            {
+                                var existingProviderConfig = ProviderConfigurations[providerConfig.Name];
+                                existingProviderConfig.Merge(providerConfig);
+                            }
+                            else
+                            {
+                                ProviderConfigurations.Add(providerConfig.Name, providerConfig);
+                            }
                         }
                         break;
                 }
             }
         }
 
+        /// <summary>
+        /// Registers given type of <typeparamref name="T"/> as bootstrap provider
+        /// </summary>
+        /// <typeparam name="T">
+        ///     Non-abstract type which implements <see cref="IBootstrapProvider"/> interface
+        /// </typeparam>
+        /// <param name="properties">
+        ///     Properties that will be passed to bootstrap provider upon initialization
+        /// </param>
+        public void RegisterBootstrapProvider<T>(IDictionary<string, string> properties = null)
+        {
+            RegisterBootstrapProvider(typeof(T), properties);
+        }
+
+        /// <summary>
+        /// Registers given type as bootstrap provider
+        /// </summary>
+        /// <param name="boostrapProviderType">
+        ///     Non-abstract type which implements <see cref="IBootstrapProvider"/> interface
+        /// </param>
+        /// <param name="properties">
+        ///     Properties that will be passed to bootstrap provider upon initialization
+        /// </param>
+        public void RegisterBootstrapProvider(Type boostrapProviderType, IDictionary<string, string> properties = null) 
+        {
+            if (boostrapProviderType == null)
+                throw new ArgumentNullException("boostrapProviderType");
+
+            if (boostrapProviderType.IsAbstract ||
+                boostrapProviderType.IsGenericType ||
+                !typeof(IBootstrapProvider).IsAssignableFrom(boostrapProviderType))
+                throw new ArgumentException("Expected non-generic, non-abstract type which implements IBootstrapProvider interface", 
+                                            "boostrapProviderType");
+
+            const string bootstrapCategoryKey = "Bootstrap";
+
+            ProviderCategoryConfiguration category;
+            ProviderConfigurations.TryGetValue(bootstrapCategoryKey, out category);
+            if (category == null)
+            {
+                category = new ProviderCategoryConfiguration()
```

Comment:
It looks like a good approach. But then we also need to make Name and Properties setters private. Otherwise, we'll still have a stinky class. But that will be a breaking change. It will broke your sample (https://github.com/gabikliot/orleans/commit/844e26b52527043135597964e41dea2aca21f19e) and might also broke any client code.